src/documentation/programming-guide.md

```diff
@@ -1013,8 +1014,186 @@ odds = results.odd
 tens = results[None]  # the undeclared main output
 ```
 
-<a name="io"></a>
-<a name="running"></a>
+## <a name="io"></a>Pipeline I/O
+
+When you create a pipeline, you often need to read data from some external source, such as a file in external data sink or a database. Likewise, you may want your pipeline to output its result data to a similar external data sink. Beam provides read and write transforms for a number of common data storage types. If you want your pipeline to read from or write to a data storage format that isn't supported by the built-in transforms, you can implement your own read and write transforms.
+
+> A guide that covers how to implement your own Beam IO transforms is in progress ([BEAM-1025](https://issues.apache.org/jira/browse/BEAM-1025)).
+
+### Reading Input Data
+
+Read transforms read data from an external source and return a `PCollection` representation of the data for use by your pipeline. You can use a read transform at any point while constructing your pipeline to create a new `PCollection`, though it will be most common at the start of your pipeline. Here are examples of two common ways to read data.
+
+#### Reading from a `Source`:
+
+```java
+// A fully-specified Read from a GCS file:
+PCollection<Integer> numbers =
+  p.apply("ReadNumbers", TextIO.Read
+   .from("gs://my_bucket/path/to/numbers-*.txt")
+   .withCoder(TextualIntegerCoder.of()));
+```
+
+```python
+pipeline | beam.io.ReadFromText('protocol://path/to/some/inputData.txt')
+```
+
+Note that many sources use the builder pattern for setting options. For additional examples, see the language-specific documentation (such as Javadoc) for each of the sources.
+
+#### Using a read transform:
+
+```java
+// This example uses JmsIO.
+PCollection<JmsRecord> output =
+    pipeline.apply(JmsIO.read()
+        .withConnectionFactory(myConnectionFactory)
+        .withQueue("my-queue"))
+```
+
+```python
+pipeline | beam.io.textio.ReadFromText('my_file_name')
+```
+
+### Writing Output Data
+
+Write transforms write the data in a `PCollection` to an external data source. You will most often use write transforms at the end of your pipeline to output your pipeline's final results. However, you can use a write transform to output a `PCollection`'s data at any point in your pipeline. Here are examples of two common ways to write data.
+
+#### Writing to a `Sink`:
+
+```java
+// This example uses XmlSink.
+pipeline.apply(Write.to(
+          XmlSink.ofRecordClass(Type.class)
+              .withRootElementName(root_element)
+              .toFilenamePrefix(output_filename)));
```

Comment:
drop altogether, just use the write transform.

Maybe use AvroIO.Write or TextIO.Write so they can be same across languages.